File: unisync_android/app/src/main/java/com/anhquan/unisync/core/plugins/run_command/RunCommandPlugin.kt
package com.anhquan.unisync.core.plugins.run_command

import android.widget.Toast
import com.anhquan.unisync.core.Device
import com.anhquan.unisync.core.plugins.UnisyncPlugin
import com.anhquan.unisync.models.DeviceMessage

class RunCommandPlugin(
    private val device: Device,
) : UnisyncPlugin(device, DeviceMessage.Type.RUN_COMMAND) {
    private object Method {
        const val EXECUTE = "execute"
    }

    fun execute(command: String) {
        sendRequest(
            Method.EXECUTE,
            mapOf(
                "command" to command
            )
        )
        Toast.makeText(
            context,
            "Executing command '$command'...",
            Toast.LENGTH_SHORT
        ).show()
    }
}

File: lib/core/plugins/run_command/run_command.plugin.dart
// ignore_for_file: non_constant_identifier_names
import 'dart:io';

import 'package:unisync/core/device.dart';
import 'package:unisync/core/device_connection.dart';
import 'package:unisync/core/plugins/base_plugin.dart';
import 'package:unisync/models/device_message/device_message.model.dart';
import 'package:unisync/utils/push_notification.dart';

class RunCommandPlugin extends UnisyncPlugin {
  RunCommandPlugin(Device device)
      : super(device, type: DeviceMessage.Type.RUN_COMMAND);
  static const _method = (EXECUTE: 'execute',);

  @override
  void onReceive(
    DeviceMessageHeader header,
    Map<String, dynamic> data,
    Payload? payload,
  ) {
    super.onReceive(header, data, payload);
    final command = data['command']!.toString().split(' ');
    final executable = command[0];
    final arguments = command.getRange(1, command.length).toList();
    Process.start(
      executable,
      arguments,
      runInShell: true,
    ).then((value) async {
      PushNotification.showNotification(
        title: 'Run command',
        text: 'Executing command "$command"...',
      );
    });
  }
}
